Add MDF Adapters

jgaff opened the pull request.

WardLT committed:
Started refactoring to run parsers via XtractHub

Commit:
Add inital MDF validation (WIP)

Some TODOs uncovered; listed in comments.

Commit:
Add inital MDF validation (WIP)

Some TODOs uncovered; listed in comments.

Commit:
Remove requirement for source_info to Validator

Validator has been modified to no longer require source_info, making MDFCS.utils unnecessary. Validator import added to TODO list.

Commit:
Remove requirement for source_info to Validator

Validator has been modified to no longer require source_info, making MDFCS.utils unnecessary. Validator import added to TODO list.

Commit:
Merge branch 'master' of https://github.com/materials-data-facility/mdf-materialsio-adapters into mdf-validation

Commit:
Add Validator

Commit:
Add Validator

Commit:
Generic MDF adapter

The GenericMDFAdapter filters out non-MDF fields by using the MDF schema (so it can always be up-to-date). This is the base for other MDF adapters.

Commit:
Generic MDF adapter

The GenericMDFAdapter filters out non-MDF fields by using the MDF schema (so it can always be up-to-date). This is the base for other MDF adapters.

Commit:
Add adapters for extractors that already output MDF format

Commit:
Add adapters for extractors that already output MDF format

Commit:
Pull MDF JSONSchema directly from Github repo instead of requiring local files

WardLT commented:
Could you spell out the features that you are adding with this PR? There's a lot going on here.

jgaff commented:
Sure. This PR adds adapters for the MDF extractors. I created a `GenericMDFAdapter` (in `generic.py`) that takes metadata containing MDF-format fields and filters out non-MDF-format fields. Since MatIO extractors are intended to always be backwards-compatible, the fields returned from an extractor can't be changed except for adding more. If more fields are added, the `GenericMDFAdapter` will automatically remove the non-MDF fields, making the metadata MDF-compliant. This structure future-proofs the MDF adapters against any backwards-compatible changes to the extractors.

However, since each extractor needs its own adapter in MatIO, I subclassed the `GenericMDFAdapter` into extractor-specific adapters (in `basic_adapters.py`). If some MDF extractors in the future require additional work to be done in the adapter (for example, to adapt an extractor that doesn't output MDF format metadata), the extractor-specific adapter would handle that, but as of now no extractors require additional processing.

This branch was created from `mdf-validation`, so it also contains the MDF Validator and changes that were made to the `__init__.py` file in that branch, if that's a source of confusion. If you'd like, I can change the target of this PR to be `mdf-validation`.

I also noticed that the `GenericMDFAdapter` requires a local copy of the MDF schema. I'll modify it to optionally pull from Github instead, for flexibility.

Commit:
Add adapter pull from remote location, simplify adapter automapping

The GenericMDFAdapter now uses new Toolbox functionality to pull from a remote schema location. The adapter also only generates the schema automapping once, in __init__, to avoid multiple requests to remote locations (this does mean that the adapter must be re-initialized if the schema changes, and the adapter must be configured via __init__ instead of in transform()

Commit:
Add adapter pull from remote location, simplify adapter automapping

The GenericMDFAdapter now uses new Toolbox functionality to pull from a remote schema location. The adapter also only generates the schema automapping once, in __init__, to avoid multiple requests to remote locations (this does mean that the adapter must be re-initialized if the schema changes, and the adapter must be configured via __init__ instead of in transform()

jgaff commented:
I've modified the `GenericMDFAdapter` to pull from remote files, including Github. It does require an update of MDF Toolbox, which is released on PyPI (v0.5.2).

@WardLT Does this look reasonable? Specifically, I can't find documentation on the expected function signature for MatIO adapters' `__init__`, and the `GenericMDFAdapter` now requires some configuration during initialization.

WardLT commented:
There is no signature for the `__init__` functions for the adapters. I've been trying to avoid needing the initializers because I'm not sure how they'd fit in to different workflow engines

I'll ponder it further

WardLT commented on mdf_matio/__init__.py:
```diff
@@ -1,16 +1,16 @@
                                           run_all_parsers, get_available_parsers)
 from mdf_matio.grouping import groupby_file, groupby_directory
-from mdf_matio.validation import validate_against_mdf_schemas
+# from mdf_matio.validation import validate_against_mdf_schemas
```
If we no longer need this import, delete it.

WardLT commented on mdf_matio/__init__.py:
```diff
@@ -103,13 +103,67 @@ def is_in_directory(f):
 
 
-def generate_search_index(directory: str, validate_records=True,
+def _call_xtracthub(data_url: str, index_options: dict, target_parsers: Set[str]) -> Iterable[ParseResult]:
```
We've changed how XtractHub will be invoked, but it doesn't look like we'll be using this `_call_xtracthub` function.

I'm not sure the right git invocation, but could you base your changes off of the `master` branch rather than the `xtracthub` branch. I do not want to merge this function into the master branch.

WardLT replied:
Sorry, I know what I'm asking will probably be a pain in the butt.

jgaff replied:
It appears that a rebase didn't have the desired effect. Is this the only change from `xtracthub` that you want removed? If so, I can just delete this function manually. Otherwise I'll need to look more into how to fully separate `adapter-class` from `xtracthub` cleanly.

WardLT replied:
Sure, feel free to delete it manually. Seems like an easy fix to another otherwise complicated dance with `git`

jgaff committed:
Delete unused import

jgaff committed:
Delete unused import

WardLT commented on mdf_matio/__init__.py:
```diff
@@ -158,11 +203,40 @@ def generate_search_index(directory: str, validate_records=True,
+                raise ValueError(rc_res["error"])
+        # Output feedstock (currently yielding)
+        yield from vald.get_finished_dataset()
+        '''
```
Please delete unused code rather than commenting it out

WardLT commented on mdf_matio/adapters/generic.py:
```diff
@@ -0,0 +1,68 @@
+class GenericMDFAdapter(BaseAdapter):
+    """Generic adapter for MDF extractors. Adapts metadata with MDF-format fields present."""
+
+    def __init__(self, schema_branch=None, schema_uri=None):
```
Why not just have it default to using the master branch for the schemas?

jgaff commented:
Unfortunately, in this case the `__init__` is needed. It would be very slow and inefficient to pull a remote schema before adapting each metadata document, especially since the schema can be easily pulled once and cached.
Whenever you have thoughts on this, I can modify  `GenericMDFAdapter.__init__` to fit the decided signature, but not having any `__init__` would be very limiting.

WardLT commented on mdf_matio/adapters/generic.py:
```diff
@@ -0,0 +1,68 @@
+                                    os.path.abspath(schema_uri),
+                                    "/" if schema_uri.endswith("/") else "")
+        # Fetch record schema with resolver
+        resolver = jsonschema.RefResolver(schema_uri, None)
```
I'd recommend creating the RefResolver as a module-level global variable. Otherwise, each parser has their own RefResolver and you will not benefit from the caching logic

WardLT replied:
That's how I approached it in the `validation.py` module

jgaff replied:
The user must be able to set the schema to validate against (e.g. `dev` or `master` branch). The earliest they can reasonably do that is in the `__init__` call for an object, because they should be able to import this class once, but validate against multiple schemas (with a separate instance for each schema). I'm not seeing any way to make the `RefResolver` a global variable without breaking some of this functionality.

WardLT replied:
Yea, let's make it a class attribute then during the `__init__` call then.

WardLT commented on mdf_matio/adapters/generic.py:
```diff
@@ -0,0 +1,68 @@
+
+
+class GenericMDFAdapter(BaseAdapter):
+    """Generic adapter for MDF extractors. Adapts metadata with MDF-format fields present."""
```
I am not sure what this means, so could you spell it out more clearly? My attempt is below:

Generic adapter for metadata produced by parsers originally created by the Materials Data Facility

The parsers originally contributed by the MDF already produce data in formats close to that required by MDF schemas.
The similarity between these parsers output and the MDF schema is therefore large enough to permit a generic adapter for all MDF-derived parsers. 
The parser functions by retrieving the fields from the metadata that are present in the MDF schema.

jgaff replied:
I changed the documentation. Do you think it's sufficient?

WardLT replied:
Yes, it looks sharp!

WardLT commented on mdf_matio/validator.py:
```diff
@@ -0,0 +1,397 @@
+
+
+class Validator:
+    """Validates MDF feedstock.
```
I like that you have a "flow" description, please spell it out in more detail.

The class documentation could also use more detail. Please explain it in a level of detail that Tyler or myself could understand what this object does and how to use it.

jgaff replied:
How does this documentation look? I've expanded it, in line with this no longer being an MDF-internal tool.

WardLT commented on mdf_matio/__init__.py:
```diff
@@ -158,11 +203,40 @@ def generate_search_index(directory: str, validate_records=True,
+        # Probably need source_id in later revision
+        source_id = dataset_metadata.get("mdf", {}).get("source_id", "unknown")
+
+        vald = Validator(schema_path=schema_path)
```
Why do you create the validator in the inner-most loop? Does that mean you create a new dataset for each set of files?

jgaff replied:
Fixed.

WardLT commented on mdf_matio/__init__.py:
```diff
@@ -158,11 +203,40 @@ def generate_search_index(directory: str, validate_records=True,
+        vald = Validator(schema_path=schema_path)
+        # Dataset validation
+        ds_res = vald.start_dataset(dataset_metadata, validation_params)
+        if not ds_res["success"]:
```
These `if not result["success"]` seem like exactly what Python's `try/except` blocks and exception system, in general, are designed for. Why not use those?

jgaff replied:
Exceptions (`ValidationError`) now used.

WardLT commented on mdf_matio/__init__.py:
```diff
@@ -158,11 +203,40 @@ def generate_search_index(directory: str, validate_records=True,
+        # Validate metadata and tweak into final MDF feedstock format
+        # Will fail if any entry fails validation - no invalid entries can be allowed
+        # TODO list:
+        #   - How should validation failures (which stop processing) be communicated?
```
I like exceptions. Maybe create a special exception type if you want them to be easily recognized by whatever uses this code.

WardLT commented on mdf_matio/__init__.py:
```diff
@@ -158,11 +203,40 @@ def generate_search_index(directory: str, validate_records=True,
+        # Will fail if any entry fails validation - no invalid entries can be allowed
+        # TODO list:
+        #   - How should validation failures (which stop processing) be communicated?
+        #   - How should the feedstock be output? (Currently yield-ed, could be written to file)
```
Maybe delegate that to another function. Keeping this function as a generator gives us flexibility to put whatever front-end we want on it.

jgaff replied:
Refactored Validator continues to `yield`, as does `generate_search_index()`.

WardLT commented on mdf_matio/__init__.py:
```diff
@@ -158,11 +203,40 @@ def generate_search_index(directory: str, validate_records=True,
+        # TODO list:
+        #   - How should validation failures (which stop processing) be communicated?
+        #   - How should the feedstock be output? (Currently yield-ed, could be written to file)
+        #   - (Future) Should feedstock be sent to Search directly here?
```
No. That's some messy token stuff, I'd imagine. This function already has enough kwargs.

jgaff replied:
This is no longer a concern; there will be a Search Ingest Action Provider that ingests entries into Search, so Xtract no longer needs to interface with Search.

WardLT commented on mdf_matio/__init__.py:
```diff
@@ -158,11 +203,40 @@ def generate_search_index(directory: str, validate_records=True,
+
+        vald = Validator(schema_path=schema_path)
+        # Dataset validation
+        ds_res = vald.start_dataset(dataset_metadata, validation_params)
```
Where do these validation params come from. Are they passed in as an argument?

jgaff replied:
Yes, the validation params will be given. They are computed by MDF, and will be part of the input to Xtract's Action Provider.

WardLT replied:
Cool. Make sure to document them in the docstring for this method then

jgaff replied:
For the record, Xtract should never need to know anything about these parameters; they will be passed directly from the Xtract input to the Validator. However, they're now documented.

WardLT commented on mdf_matio/adapters/basic_adapters.py:
```diff
@@ -0,0 +1,60 @@
+    pass
+
+
+class JSONAdapter(GenericMDFAdapter):
```
Is this different than the existing JSON parser? If so, let's pick which one we want to keep

jgaff replied:
I believe it is different, yes. The `JSONAdapter` is the adapter for the MDF-contributed `JSONParser` in MatIO. The existing code seems to also be an extractor? I'm unsure what purpose it serves, but I believe the case of extracting and adapting JSON to be served by the `JSONParser` and `JSONAdapter`.

WardLT replied:
I think we've sufficiently discussed this one :smiley:

WardLT commented on mdf_matio/validator.py:
```diff
@@ -0,0 +1,397 @@
+        # Validate required fields
+        # TODO: How should this validation be done?
+        # The metadata conforms to the schema, there are just extra
+        # `requires` values. Perhaps add these to the schema instead?
```
I support this. What we have here seems like a challenge that is solved by the JSON schema and it's Python validator.

jgaff replied:
It's also a challenge to correctly monkeypatch the schema. For context, the extra "requires" values are dynamic, and set by the user (or, more commonly, set by options chosen by the user) during dataset submission. It's not possible to bake these "requires" into the base schema; they would have to be added at runtime.
Do you have thoughts on how to easily add the extra "requires" to the correct places in the preset JSONSchema? (I've been bitten trying to do that before; JSONSchema is just verbose enough that it's difficult to programmatically traverse dynamically, without errors.)

WardLT replied:
Ah, thanks for clarifying! 

No, I don't have any ideas on how to traverse the schema. Maybe you could write out JMESPath specs on the fields you want to check and delegate recursing into the records to that code? 

That said, this works and you've already marked this as a "semi-temporary" measure

jgaff replied:
Using JMESPath could be an enhancement in the future, but for now since this code is functional I think we can punt it down the road.

WardLT commented on mdf_matio/validator.py:
```diff
@@ -0,0 +1,397 @@
+                        missing.append(field_path)
+                        break
+            if missing:
+                return {
```
It should be OK to throw exceptions instead, given how you have "if not success then raise exception" to catch these "exceptions." Not using exceptions masks that the error actually came from here in a stacktrace.

WardLT commented on mdf_matio/validator.py:
```diff
@@ -0,0 +1,397 @@
+        # Load schema
+        with open(os.path.join(self.__schema_dir, "record.json")) as schema_file:
+            schema = json.load(schema_file)
+        resolver = jsonschema.RefResolver(base_uri="file://{}/".format(self.__schema_dir),
```
Make your ref-resolver a module-level global variable or, at least, a class attribute of the validator. They do cache schemas but this the cache dies with this object.

jgaff replied:
This is resolved with the `mdf-validation` merge. The `RefResolver` is now created in the `__init__`. Making it a module-level global would run into the same issues as the `GenericMDFAdapter` above.

WardLT commented on mdf_matio/validator.py:
```diff
@@ -0,0 +1,397 @@
+
+        # BLOCK: material
+        # elements
+        if rc_md["material"].get("composition"):
```
This code seems like a candidate for breaking out into a separate function.

jgaff replied:
Why is that? It's not reused anywhere, and is (reasonably) simple. I'm aware there are several commented-out blocks, that have been previous approaches I want to keep around to turn back on if necessary.

WardLT replied:
My reasoning would be to make it easier to test by being able to invoke it in unit tests without engaging all the other parts of this code. That and you could document the desired behavior of the code in a place where users are more likely to find it.

jgaff replied:
It's easy enough to have tests that engage this code (records with `material.composition` specified); I'm not convinced there's a significant testing benefit. Additionally, users should not be using this code directly anyway.

WardLT commented on mdf_matio/validator.py:
```diff
@@ -0,0 +1,397 @@
+        try:
+            jsonschema.validate(rc_md, schema, resolver=resolver)
+        except jsonschema.ValidationError as e:
+            return {
```
As above, use exceptions not this custom implementation of try/catch blocks.

WardLT commented on mdf_matio/validator.py:
```diff
@@ -0,0 +1,397 @@
+
+        # Write out to file
+        json.dump(rc_md, self.__tempfile)
+        self.__tempfile.write("\n")
```
If we do not perform any further computation using a record after this point, why store it at all? Could we just return the validated records and let the user of the class decide if it is needed to store them or whether they can be passed out of the generator.

WardLT replied:
[Discussed in main thread]

WardLT commented on mdf_matio/validator.py:
```diff
@@ -0,0 +1,397 @@
+        self.__dataset = None
+        return
+
+    def status(self):
```
If you agree this class need not hold on to the records, then we can kill this function and `get_finished_dataset`

WardLT commented on mdf_matio/adapters/basic_adapters.py:
```diff
@@ -0,0 +1,60 @@
+    pass
+
+
+class ImageAdapter(GenericMDFAdapter):
```
We already have a image adapters listed as a "no op" parser. Do we need this?

WardLT replied:
Same for generic

jgaff replied:
Yes, this is better than a "no-op" adapter. The `GenericMDFAdapter` provides forward-compatibility, because MatIO extractors' output can only change by adding additional fields. Therefore, if all non-MDF-format fields are removed (by the `GenericMDFAdapter`), the adapters don't need to be changed if the extractor adds additional fields.

WardLT replied:
Great, then could you remove the uses of `noop` parser? It seems like there isn't a reason we should use that one in place of this GenericMDFAdapter

WardLT commented on mdf_matio/adapters/basic_adapters.py:
```diff
@@ -0,0 +1,60 @@
+    pass
+
+
+class FilenameAdapter(GenericMDFAdapter):
```
What is the filename parser? Is it different than the "generic parser"

jgaff replied:
Yes, the `FilenameParser` extracts metadata from a filename (not the contents of a file). I can't remember if this made it into MatIO yet, or if there were issues, but a filename extractor exists in MDF right now, separate from the file info extractor (which extracts file size, type, path, etc.).

WardLT replied:
Cool!

WardLT commented on mdf_matio/__init__.py:
```diff
@@ -158,11 +203,40 @@ def generate_search_index(directory: str, validate_records=True,
+            if not rc_res["success"]:
+                raise ValueError(rc_res["error"])
+        # Output feedstock (currently yielding)
+        yield from vald.get_finished_dataset()
```
It appears that each record is treated independently by the validator. If true, we don't need to store the metadata on disk and could just yield the record after validating / adding the dataset-level metadata.

WardLT commented on setup.py:
```diff
@@ -11,7 +11,7 @@
+    install_requires=['pypif_sdk', 'jmespath>=0.9.4', 'jsonschema>3', 'mdf_toolbox>=0.5.2'],
     include_package_data=True,
     entry_points={
         'materialsio.adapter': [
```
Eventually, we need to add in entry_points for the adapters so that MaterialsIO can find them.

WardLT reviewed (changes requested):
Let's talk about the `__init__` function after we resolve these issues.

Short version: 
- The Validator class holds on to too much state, which makes it unnecessarily complex to implement or use. Big issue: Once a record is done validating it, you need not hold on to it and can pass it back to the user of the class to deal with.
- The GenericMDFAdapter looks good. It just needs some more documenting and we need to de-duplicate some of the derivative adapters with those already present in this library (e.g., Image and JSON)

jgaff commented:
To give you some context on the Validator, it's designed the way it is for some requirements MDF Connect has, that now aren't going to be necessary for Xtract. For example, the records are not returned immediately after validation because the feedstock format Connect expects contains the dataset entry first, and then all the record entries. This allows Connect to access only the dataset entry of the feedstock. However, the dataset entry can't be returned until all the records are processed, because of some features Ben wants/has wanted (computing the total size of the dataset by the sizes of the records, as an example).

Since the Xtract flow is going to be architecturally different from the current Connect flow, I can revisit these requirements, and I agree that the Validator can just send each record entry back upon a successful validation, so I'll implement that.

As for the `GenericMDFAdapter`, I'll add additional documentation about its purpose and use. I think it supersedes the existing adapters (although the existing CSV/JSON "adapters" look like extractors to me, requiring a user mapping), because the `GenericMDFAdapter` will automatically update with any changes to the MDF schema, and handles any metadata document that has MDF schema fields. I think all MDF adapters should subclass the `GenericMDFAdapter`, and use its `transform()` to filter out non-MDF metadata.

The other issue you raise in a few spots is the use of exceptions vs. returning a `"success": False`. While I understand returning `success` is less Pythonic, I've found it useful to have a clear distinction between an error state (exception) and an unsuccessful result. I understand your point, though, and I'll change the behavior to throw more exceptions in the interest of easy interoperability.

Commit:
Merge branch 'adapter-class' of https://github.com/materials-data-facility/mdf-materialsio-adapters into adapter-class

jgaff commented:
I'm attempting to rebase `adapter-class` onto `master` instead of `xtracthub`. I _think_ `rebase` is the correct command for this, but we may have to rollback if I screw it up.

jgaff committed:
Merge pull request #4 from materials-data-facility/mdf-validation

Merge modified Validator into adapter-class

WardLT commented:
Thanks for the detailed reply, Jonathon. I think I'm on the same page about the purpose of Validator and I'm on board with your plan for how to more forward. Same with the MDFGeneralAdapter.

Wrt "csv" and "json" adapters, it true that they kind of blur the line between adapter and parser. Ex: 'A JSON file is already in a JSON format, why are you talking about "extracting it into a JSON format"?' I think of the split as "the extractor" gets the data off disk and adapter reshapes it into specific JSON format. We have a JSON adapter because it takes user-provide context to shape arbitrary JSON into a MDF-compatible format. 

Correct me if wrong, but the MDFGenericAdapter for JSON does not support taking user-specified mappings and my version does not support taking JSONs that are already in MDF format. Do you think we can merge those two?

jgaff commented:
For the JSON extractor in particular, it does seem like the extractor/adapter line is a little fuzzy. I mentally modelled it along the lines of the current MDF flow, which doesn't have any adapters but does extract JSON. If you think it will be more useful to have the `JSONAdapter` perform the "take JSON + mapping and output different JSON" I can make that change (I'm not going to change the `GenericMDFAdapter` to take user mappings, because that's not its role, but the `JSONAdapter` easily could and would make sense).

However, this doesn't really hold well for the CSV extraction. CSV isn't in JSON format, so it could be "extracted" into JSON format. Exactly what this would look like may not be totally clear, though. How do you think we should handle CSV files?

Commit:
Address easy PR feedback.

- Delete unused code
- Default GenericMDFAdapter to master branch of schema
- Additional GenericMDFAdapter documentation
- Create Validator earlier in generate_search_index()

WardLT commented:
- _JSON_: Shall we keep the two JSON adapters, but give them different names? Maybe `MDFJson` for files that are anticipated to already be complaint with the MDF schema?

- _CSV_: The CSV extractor works by loading the data from disk into individual JSON records ([link](https://materialsio.readthedocs.io/en/latest/_modules/materials_io/csv.html#CSVParser)). The adapter is pretty similar to the JSON adapter. It applies a mapping to generate a list of MDF-compliant JSON objects.

jgaff commented:
### JSON

> keep the two JSON adapters

I disagree with having multiple "JSON Adapters" for MDF. Either we have the JSONExtractor do the mapping work and the JSONAdapter only filters non-MDF fields, or the JSONAdapter does both, but having two separate JSONAdapters seems like it's only going to be more confusing and less useful.

However, we should define what we mean here. I see three separate functions:
1) `materials_io.json.JSONExtractor`, (the "JSONExtractor") which is a copy of the MDF JSON extractor. This function takes a JSON document and a mapping of `output_field: input_field`, and extracts accordingly.
2) `mdf_matio.adapters.basic_adapters.JSONAdapter` (the "MDF JSONAdapter") is the adapter to match the JSONExtractor. It is not particularly useful, as output from the Extractor should be in MDF format, but MatIO requires an adapter for every extractor.
3) `mdf_matio.adapters.mappable.JSONAdapter` (the "Parsing JSONAdapter"), which appears to do the same thing as the JSONExtractor, but is in `mdf_matio` and uses `JMESPath`.

I can merge 1 and 2, to create a new "MDF JSONAdapter" that takes a JSON document and mapping, extracts accordingly, and filters out any non-MDF fields (which is what 1 and 2 do now, but separately). This has some implications that we should consider:
- There will be an adapter in mdf_matio with no corresponding extractor in MatIO
- JSON files cannot be processed using MatIO
- For the MDF case, Xtract will have to keep track not just of extractor/adapter pairs to run, but also this one-off adapter

I don't know what to do with 3, because it's essentially a duplicate of 1.



### CSV

Is your proposal then to keep the CSVExtractor as-is (requiring a mapping to MDF format), so we can then keep the CSVAdapter the same (just a subclass of the `GenericMDFAdapter`)? I'm okay with that plan, and it works for the other mappable types as well (YAML, XML, etc.).
Otherwise, I'm not sure what your idea is.

WardLT commented:
For clarity, did you mean to say the JSONExtractor currently has or that you envision it having the ability to take a mapping as context?

jgaff commented:
The existing JSONExtractor in MaterialsIO ([link](https://github.com/materials-data-facility/MaterialsIO/blob/mdf-extractors/materials_io/json.py)) I ported over from MDF has that function.

WardLT commented:
Oh! I thought you were referring to the one in this library, which does not, despite the fact that you did list the full path :man_facepalming:

WardLT commented:
The [JSONExtractor in this library](https://github.com/materials-data-facility/mdf-materialsio-adapters/blob/master/mdf_matio/parsers/json.py) does not require a mapping. I intentionally delegate the "mapping" to the adapter class because "mapping" is a transformation into MDF format rather than a "just give me what this file has."

In the light of this confusion over whose JSONExtractor we meant, my proposal would be to drop the matio version of the extractor (1)  or denote that it is explicitly a extractor that reads JSON files that are already in MDF format. 

I am OK with there being a differentiation between "MDF-formatted" JSON files and arbitrary JSON, and expecting that there would be different "extractor+adapter" pairs for each type.

jgaff commented:
Let's call `mdf_matio.parsers.json.JSONExtractor` (4), then. I don't really understand why (4) exists, why it's in `mdf_matio` instead of MatIO, or what it does except detect line-delimited JSON files. It seems like (4) + (3) roughly has the same goal as (1) + (2), delegated differently, but that makes (4) the function with little real purpose instead of (2).


I do think there is some context you are missing. (1) is not actually MDF-specific. It should not be noted as "explicitly an extractor that reads JSON files that are already in MDF format" because it does not do that. (1) takes a JSON file and a mapping, and outputs a JSON document translated by the mapping. It does not touch MDF's schemas at all. If I understand the goals of MatIO, this is the correct approach. (1) will translate _any_ JSON format into _any_ JSON format, but then (2) forces the JSON into MDF format. (1) "extracts" well-behaved JSON out of messy/nonstandard JSON (given the appropriate `context`), which seems like an Extractor behavior, and (2) takes the output of an Extractor and turns it into a user-specific format, which seems like an Adapter behavior.

I disagree with there being an "MDF JSON" pipeline and a "regular JSON" pipeline because both cases can be served with (1) and (2). If a JSON file submitted to MDF is not in MDF format, (1) maps it to MDF format and (2) cleans out non-MDF fields. If a JSON file in MDF format is submitted (which would be unexpected), (1) will map it to itself and (2) will clear out nothing. In the latter case, it may not be useful to run (1) and (2), but it works fine and is an unlikely edge case.

I can support merging (1) and (2) into a super-(2), which drops (1) from MatIO, but raises the implications I mentioned earlier. I also propose we drop (3) and (4) because they are redundant.

WardLT commented:
Alright, I'm sold :) Let's go with keeping your pipeline as the JSON route (1+2), and drop all other routes.

Since you already have a parser that applies mappings and an adapter that removes excess fields, let's keep that pattern. I agree that they could all be completed in the same class, but also note that such a merging would break down the separation of MDF-specific logic (e.g., care about the schema).

Also, could you make a PR for your modifications to MatIO? I didn't realize this branch required modifications to matio to function. 

I also forgot to your comment about CSVs. Sure, keeping it as-is and making mapping mandatory is OK with me. We can discuss any technical details on the CSV parser (e.g., whether it needs to return a `dict`) when you issue a PR to materialsIO.

Further thoughts below. 

------

> I don't really understand why (4) exists, why it's in mdf_matio instead of MatIO, or what it does except detect line-delimited JSON files.

Indeed, it only detects JSON-LD formated JSON files and should be in MatIO. I'm not sure what motivated me to put it in this library

------

The only capabilities not included in your JSON extractor vs mine are the ability to read JSON-LD files and JMESPath-based mappings. Shall we eventually consider upgrading your Extractor to include that and JMESPath-based mapping specifications?

WardLT commented on mdf_matio/adapters/generic.py:
```diff
@@ -0,0 +1,83 @@
+    reduces the maintenance burden for MDF adapters.
+
+    All MDF adapters should subclass this adapter, and should call this adapter's transform()
+    to gain forwards-compatibility.
```
Nice! This docs are wonderful

WardLT reviewed (commented):
I made a few more comments. I don't think there's anything in here we haven't discussed yet.

Thanks for your patience!

jgaff commented:
I'll remove (3) and (4) from mdf_matio, then, and create a PR in MatIO with additional discussion topics.

JMESPath seems a little overcomplicated for this application (and confusing for users), but if it's useful we can add support for it in the JSONExtractor (1) later on.

WardLT commented:
Sounds awesome. Thanks, @jgaff !

I agree that JMESPath is confusing and that we should only roll it out if/when needed. Luckily, the "a.b.c" notation we use in our current mapping specs has the same meaning in JMES.

Commit:
Refactor Validator

Validator now has only one user-facing method, validate_mdf_dataset(), which is a generator that accepts unvalidated records and yields validated records (more details in the docstrings). Additionally, the Validator now throws a ValidationError on any error instead of returning a failure status.

Commit:
Delete _call_xtracthub and related imports

Commit:
Slightly clarify Validator documentation

Commit:
Delete redundant JSONParser and JSONAdapter

Commit:
Modify __init__.generate_search_index() to match updated Validator

Todo: Add additional arguments to generate_search_index() to take other required parameters, in whatever form is convenient for Xtract

Commit:
Remove noop extractor/parser references; should use GenericMDFAdapter instead

Commit:
Move FileAdapter out of __init__.py

Commit:
Add entry points for adapters

Commit:
Fix importing old Validator name

Validator was renamed MDFValidator

WardLT commented on mdf_matio/__init__.py:
```diff
@@ -120,36 +120,42 @@ def generate_search_index(directory: str, validate_records=True,
-    parse_results = run_all_parsers(directory, include_parsers=list(target_parsers),
-                                    adapter_map='match', parser_context=index_options,
-                                    adapter_context=index_options)
+    # NOTE: No longer using _call_xtracthub()
```
Could you replace this commented-out code with the original call to `run_all_parsers`?

WardLT commented on mdf_matio/validator.py:
```diff
@@ -0,0 +1,362 @@
+        # Process all records the user has
+        while record is not None:
+            record = yield self._validate_record(record)
+        # Yield once more to avoid forcing user to catch StopIteration
```
This is pretty cool! I've never seen the `send` function before and I think it's a great use of it!

WardLT reviewed (commented):
Looks great! Would you mind fixing up the tests/flake8 issues?

Then, from my view, this is ready to merge. Cool stuff!

Commit:
Flake8 cleanup

jgaff commented:
I've fixed the Flake8 failures (mostly lines too long in existing code), but the `generate_search_index()` code isn't passing the tests. It looks like a `None` is being operated on as if it's a dict, but I don't know what the intended behavior should be. How should that be addressed?

WardLT commented:
It looks like I introduced `parse_config` as an eventual route for supporting merging metadata for all files in the same directory. 

If we do not want to support directory merging yet, perhaps remove `parse_config` and the lines where we merge directories (lines 142-146)?

jgaff commented:
> If we do not want to support directory merging yet

We definitely do want to support group-by-directory. However, I think Tyler has tackled that case. Is the `parse_config` important for his implementation (for that matter, is he actually using `generate_search_index()` for more than an example)?

WardLT commented:
@tskluzac do you want "merge by directory"?

Commit:
Update GenericFileAdapter

Expected to fail tests until MDF Toolbox 0.5.3 is published

jgaff commented:
@WardLT I talked to Tyler and he said he's not directly using `generate_search_index()` as a callable function, he's just using it as an example flow. Since the failing tests are failing on `generate_search_index()` (because it isn't configured with real, usable data), should we just move forward with the merge?

WardLT committed:
Mark search index tests as expected failures

We'll deal with whether they should state at all later.

WardLT commented:
Agreed, let's go ahead and merge without fixing the tests. We'll decide on whether we should keep the `generate_search_index` function later.

I've added some `xfail` marks for those tests, and will merge once I get the tests passing.
